### PR TITLE
Bluetooth: controller: split: Remove use of k_cpu_idle

### PR DESCRIPTION
**Pull request by cvinayak**
Remove use of k_cpu_idle in controller, and refactor the
implementation used to start, wait and stop clocks needed
by the controller.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

**Comment by cvinayak** [comment]
CC @stephanosio @kruithofa

**Review by George-Stefan — APPROVED** [review]
As I'm not familiar with the fine points of the Nordic sw/hw, I did a quick review.

**Review by dagbja — APPROVED** [review]

**Review by thoh-ot — COMMENTED** [review]
Why is conn_release inline in ull_master.c but not in ull_adv.c?

**Review by carlescufi — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Comment by carlescufi** [comment]
@cvinayak there are some CI issues that need fixing

File: subsys/bluetooth/controller/ll_sw/ull_adv.c
Comment (kruithofa) [inline subsys/bluetooth/controller/ll_sw/ull_adv.c]: In the current implementation(s) lll_clock_wait always returns 0.
If this is for preparing for an implementation where we actually can check for stable clock I propose saying this in a comment

**Review by kruithofa — APPROVED** [review]
Only minor comments; nothing that needs fixing before my approval

**Comment by kruithofa** [comment]
> Why is conn_release inline in ull_master.c but not in ull_adv.c?

I propose removing the inline in ull_master.c; depending on compiler (and compiler flags) the function may or may not be inlined. In my test-setup conn_release was kept as a separate function (in ull_master.c) both with and without the inline keyword

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c
Comment (nordic-krch) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c]: can you add assert when in interrupt?

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c
Comment (nordic-krch) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c]: can you add assert when in irq context?

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c]: Damn, I am in IRQ, I cant use semaphores to wait :-(

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c]: @ioannisg I cant use k_sem_take inside ISR. Any alternative ideas to replace k_cpu_idle?

Comment (nordic-krch) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c]: isn't it used only in the test? Maybe you could handle it differently? `lll_hfclock_on(cb)`

**Comment by cvinayak** [comment]
@thoh-ot @kruithofa 
> 
> > Why is conn_release inline in ull_master.c but not in ull_adv.c?
> 
> I propose removing the inline in ull_master.c; depending on compiler (and compiler flags) the function may or may not be inlined. In my test-setup conn_release was kept as a separate function (in ull_master.c) both with and without the inline keyword

conn_release in ull_master.c is called only in one place, and I want to provide the compiler a hint to inline it. Where as in ull_adv.c its conn_release is called in two places.

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c]: yes, will add it. The lll_clock_wait and lll_hfclock_on_wait shall only be called in thread.

Comment (ioannisg) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c]: So: my "problem" with k_cpu_idle() is that it is "specified" to be used only in the Zephyr idle thread. Alternatively, you can _keep_ using the arch_cpu_idle _directly_ which is the ARM-implementation of k_cpu_idle, as long as you understand the consequences of it. Please, check https://github.com/zephyrproject-rtos/zephyr/pull/23511 for the details

Comment (ioannisg) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c]: > Damn, I am in IRQ, I cant use semaphores to wait :-(

@cvinayak So the code is really supposed to busy-wait inside an ISR?

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c]: @ioannisg no, it was my fault, only the call inside thread needs waiting. Calls from ISR do not need to wait, prepare will start and give enough time until event for the clock to settle. I will add assert checks and this PR (original commits) should hold good.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c]: @nordic-krch k_sem_take already has assert check on arch_is_in_isr. Hence, I dont need to add in controller.

**Review by nordic-krch — APPROVED** [review]

**Comment by cvinayak** [comment]
Added DNM, so that I can replace K_FOREVER with a suitable timeout value.

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c]: Observed a settling time from power up of ~280ms.

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c]: Added.

File: subsys/bluetooth/controller/ll_sw/ull_adv.c
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv.c]: Updated with timeout in k_sem_take and returns the err value.

**Pull request merged**
